tools/refguide_check.py

```diff
@@ -0,0 +1,952 @@
+#!/usr/bin/env python
+"""
+refguide_check.py [OPTIONS] [-- ARGS]
+
+Check for a NumPy submodule whether the objects in its __all__ dict
+correspond to the objects included in the reference guide.
+
+Example of usage::
+
+    $ python refguide_check.py optimize
+
+Note that this is a helper script to be able to check if things are missing;
+the output of this script does need to be checked manually.  In some cases
+objects are left out of the refguide for a good reason (it's an alias of
+another function, or deprecated, or ...)
+
+Another use of this helper script is to check validity of code samples
+in docstrings. This is different from doctesting [we do not aim to have
+numpy docstrings doctestable!], this is just to make sure that code in
+docstrings is valid python::
+
+    $ python refguide_check.py --doctests optimize
+
+"""
+from __future__ import print_function
+
+import sys
+import os
+import re
+import copy
+import inspect
+import warnings
+import doctest
+import tempfile
+import io
+import docutils.core
+from docutils.parsers.rst import directives
+import shutil
+import glob
+from doctest import NORMALIZE_WHITESPACE, ELLIPSIS, IGNORE_EXCEPTION_DETAIL
+from argparse import ArgumentParser
+from pkg_resources import parse_version
+
+import sphinx
+import numpy as np
+
+sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'doc', 'sphinxext'))
+from numpydoc.docscrape_sphinx import get_doc_object
+
+if parse_version(sphinx.__version__) >= parse_version('1.5'):
+    # Enable specific Sphinx directives
+    from sphinx.directives import SeeAlso, Only
+    directives.register_directive('seealso', SeeAlso)
+    directives.register_directive('only', Only)
+else:
+    # Remove sphinx directives that don't run without Sphinx environment.
+    # Sphinx < 1.5 installs all directives on import...
+    directives._directives.pop('versionadded', None)
+    directives._directives.pop('versionchanged', None)
+    directives._directives.pop('moduleauthor', None)
+    directives._directives.pop('sectionauthor', None)
+    directives._directives.pop('codeauthor', None)
+    directives._directives.pop('toctree', None)
+
+
+BASE_MODULE = "numpy"
+
+PUBLIC_SUBMODULES = [
```

Comment:
It is important to check that this list of public NumPy submodules is reasonably complete. I just noticed that `lib.recfunctions` was previously absent, so refguide_check wasn't executing docstrings in that module, some of which were the subject of reviewer comments (probably diffs originated from before refguide switch when I was using vanilla doctests).

Also, even if I add `lib.recfunctions`, not all function docstrings therein are executed -- it seems I also need to add functions to `__all__` in the module, but I'm hesitant do start doing that more broadly without reviewer approval first. Thoughts on this?

Reply:
I think skipping private functions, which should be everything not in `__all__` with, or without the underscore, would probably be OK.  If they are not intended to be private then they should be added. That said, `lib.recfunctions` has always been an oddball, as it is not imported at the top level.

@shoyer Not all functions in `lib.recfunctions` have dispatchers.

Reply:
@charris looking through `lib.recfunctions` the functions without dispatchers seem to not take arrays as inputs (e.g., `get_fieldspec`). So these properly don't have dispatchers -- there's no array argument on which to call a dispatcher method (we don't have duck-typed dtypes). Let me know if there's anything in particular that you think I missed, though...